compiler/rustc_mir_build/src/build/expr/stmt.rs

```diff
@@ -140,15 +138,15 @@ impl<'a, 'tcx> Builder<'a, 'tcx> {
                 // it is usually better to focus on `the_value` rather
                 // than the entirety of block(s) surrounding it.
                 let adjusted_span = (|| {
-                    if let ExprKind::Block { body } = expr.kind {
+                    if let ExprKind::Block { body } = &expr.kind {
                         if let Some(tail_expr) = &body.expr {
-                            let mut expr = tail_expr;
-                            while let rustc_hir::ExprKind::Block(subblock, _label) = &expr.kind {
-                                if let Some(subtail_expr) = &subblock.expr {
-                                    expr = subtail_expr
-                                } else {
-                                    break;
-                                }
+                            let mut expr = &*tail_expr;
+                            while let ExprKind::Block {
+                                body: Block { expr: Some(nested_expr), .. },
+                            }
+                            | ExprKind::Scope { value: nested_expr, .. } = &expr.kind
```

Comment:
Here I only took the addition of `Scope`s into account, are there any `Adjustment`s that could mess this up?

Reply:
Possibly `ValueTypeAscription`, but I'm not sure how to even write a test for that. The current version seems fine to me, we can fix the diagnostics later

cc @rust-lang/wg-diagnostics